server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4507,4 +4503,18 @@ public void waitForPrimaryTermAndGeneration(long primaryTerm, long segmentGenera
         );
     }
 
+    /**
+     * Ensures that the shard is ready to perform mutable operations.
+     * This method is particularly useful when the shard initializes its internal
+     * {@link org.elasticsearch.index.engine.Engine} lazily, as it may take some time before becoming mutable.
+     *
+     * The provided listener will be notified once the shard is ready for mutating operations.
+     *
+     * @param listener the listener to be notified when the shard is mutable
+     */
+    public void ensureMutable(ActionListener<Void> listener) {
+        indexEventListener.beforeIndexShardMutableOperation(this, listener.delegateFailure((l, unused) -> {
+            // TODO: Acquire ref to engine and retry if it's immutable again?
```

Comment:
Don't we need to notify `l.onResponse(null)` here? I wonder if I'm confused, else I don't see how it works since we only delegate failures with the current code.

Reply:
🤦 I wanted to illustrate where I think that we should acquire the ref and messed up... thanks for noticing! Fixed in https://github.com/elastic/elasticsearch/pull/122241/commits/5b6aa8731f82ae61347ac2d96ea551991688c5c9

Reply:
Thanks! Is the `ref` you mentioned here the one that ES-10826 wants to do? If it's that, then maybe add the ticket number to the TODO as well.

Reply:
Yes, that's my idea. Added a reference to the ticket in https://github.com/elastic/elasticsearch/pull/122241/commits/61aa6ed10efdd924fee5475d59dbcf93c37f7400